bakery-js/src/minidom.ts

```diff
@@ -0,0 +1,337 @@
+  epub: 'http://www.idpf.org/2007/ops',
+  ncx: 'http://www.daisy.org/z3986/2005/ncx/',
+  opf: 'http://www.idpf.org/2007/opf',
+  dc: 'http://purl.org/dc/elements/1.1/',
+  books: 'https://openstax.org/namespaces/book-container',
+  cont: 'urn:oasis:names:tc:opendocument:xmlns:container',
+}
+
+const xpathSelect = useNamespaces(NAMESPACES)
+
+// Custom element attributes go here.
+// If this becomes annoying then just set IntrinsicElements = any
+declare global {
+  namespace JSX {
+    interface IntrinsicElements {
+      'cont:container': { version: '1.0' }
+      'cont:rootfiles': {}
+      'cont:rootfile': {
+        'media-type': 'application/oebps-package+xml'
+        'full-path': string
+      }
+
+      'opf:package': { version: '3.0'; 'unique-identifier': string }
+      'opf:metadata': {}
+      'opf:meta': {
+        property: 'dcterms:modified' | 'dcterms:license' | 'dcterms:alternative'
+      }
+      'opf:manifest': {}
+      'opf:item': {
+        id: string
+        'media-type': string
+        href: string
+        properties?: string
+      }
+      'dc:title': {}
+      'dc:language': {}
+      'dc:identifier': { id: string }
+      'dc:creator': {}
+      'opf:spine': any
+      'opf:itemref': any
+
+      // Schema: https://github.com/w3c/epubcheck/blob/main/src/main/resources/com/adobe/epubcheck/schema/20/rng/ncx.rng
+      'ncx:ncx': { version: '2005-1' }
+      'ncx:head': {}
+      'ncx:meta': { name: string; content: string | number }
+      'ncx:docTitle': {}
+      'ncx:text': {}
+      'ncx:navMap': {}
+      'ncx:navPoint': { id: string }
+      'ncx:navLabel': {}
+      'ncx:content': { src: string }
+
+      'h:title': {}
+      'h:link': { rel: 'stylesheet'; type: 'text/css'; href: string }
+    }
+  }
+}
+
+export type JSXNode = {
+  tagName: string
+  config: AttrsOrChildren
+  source: SourceInfo
+}
+type AttrsOrChildren = { [key: string]: string | undefined } & {
+  children?: JSXNode | Array<JSXNode>
+}
+type SourceInfo = {
+  fileName: string
+  lineNumber: 0
+  columnNumber: 0
+}
+// See https://www.typescriptlang.org/tsconfig#jsx for more info
+export function jsxDEV(
+  tagName: string,
+  config: AttrsOrChildren,
+  _1: undefined,
+  _2: boolean,
+  source: SourceInfo
+): JSXNode {
+  return { tagName, config, source }
+}
+
+export function fromJSX(j: JSXNode) {
+  const [tagName, prefix] = j.tagName.split(':').reverse()
+  const { children, ...attrs } = j.config
+  const source = {
+    ...j.source,
+    source: { fileName: j.source.fileName, content: null },
+  }
+
+  const doc = parseXml(
+    `<${tagName} xmlns="${assertValue(
+      (NAMESPACES as any)[prefix],
+      `Unknown namespace prefix '${prefix}'`
+    )}"/>`
+  )
+  setPos(doc.documentElement, source)
+
+  const $doc = dom(doc)
+  const $root = dom($doc.doc.documentElement)
+  $root.attrs = attrs
+  $root.children =
+    children === undefined
+      ? []
+      : Array.isArray(children)
+      ? children.map((c) => $doc.fromJSX(c))
```

Comment:
This could probably use additional indentation.

Reply:
Yep, after debating how to format the code we opted to just use a code formatter. This reminded me that I need to set it up to run in CI (and maybe as a commit hook)